pkg/storagev2/transaction.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package storage
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// ErrTxDone is returned by any opCode that is performed on
+// a transaction that has already been committed or rolled back.
+var ErrTxDone = errors.New("storage: transaction has already been committed or rolled back")
+
+// Tx represents an in-progress Store transaction.
+// A transaction must end with a call to Commit or Rollback.
+type Tx interface {
+	// Commit commits the transaction.
+	Commit() error
+
+	// Rollback aborts the transaction.
+	Rollback() error
+}
+
+// TxStore represents a Tx Store where all
+// operations are completed in a transaction.
+type TxStore interface {
+	Tx
+	Store
+
+	NewTx(*TxState) TxStore
```

Comment:
So I think the `NewTx` function should not be here. The problem being we already have the object implementing Tx and Store. If users dont use NewTx, this could cause panics as the TxState is not initialised. In any case, self instantiating new objects seems not good.